docs/dependencies-node-roadmap.md

```diff
@@ -0,0 +1,104 @@
+
+## Generating Dependencies
+
+As the project changes due to evaluations and design-time builds CPS pushes changes through TPL dataflow blocks. The [`DependencyRulesSubscriber`][16] (implementing [`IDependencyCrossTargetSubscriber`][11] via [`CrossTargetRuleSubscriberBase<T>`][17]) receives this data. The various implementations of [`ICrossTargetRuleHandler<T>`][18] both specify what rules to listen for and handle processing changes to the associated items--see the `HandleAsync` methods in both [`CrossTargetRuleSubscriberBase<T>`][17] and [`DependenciesRuleHandlerBase`][19].
+
+These handlers add the added and removed items to a [`DependenciesRuleChangeContext`][20]. This fires its `DependenciesChanged` event and passes along the [`DependenciesRuleChangeContext`][20] to event subscribers.
+
+Each project has an instance of [`DependencySubscriptionsHost`][7] implementing [`IDependenciesSnapshotProvider`][2]. It subscribes to this `DependenciesChanged` event, ultimately handling it in `UpdateDependenciesSnapshotAsync`. If there are relevant changes it creates a new [`DependenciesSnapshot`][21] and fires its own `SnapshotChanged` event.
+
+The [`DependenciesProjectTreeProvider`][1] subscribes to this event and handles updating the tree.
+
+## Translating snapshots to trees
+
+Most of the work of translating [`IDependency`][4]s to `IProjectTree`s is done by [`IDependenciesTreeViewProvider`][3] and its singular implementation, [`GroupedByTargetTreeViewProvider`][8]. It takes an [`IDependenciesSnapshot`][6] and generates the nodes for the target frameworks, the groupings under each framework, (Assemblies, Analyzers, Packages, Projects, etc.), and the top-level nodes under each of those groupings. In the common case that a project has a single target framework it leaves out the framework node entirely and simply hangs the different groupings directly off the `IProjectTree` for the Dependencies node.
+
+> Aside: While there is currently only one implementation of [`IDependenciesTreeViewProvider`][3] you could imagine others that organize the nodes in different ways: flattening the top-level items into a single list or organizing nodes first by type then by target framework. Should we create other implementations we would also need some way to switch between them; there is currently no support for doing so.
+
+The [`GroupedByTargetTreeViewProvider`][8] traverses down the existing `IProjectTree` and the new [`IDependenciesSnapshot`][6] in parallel starting from the Dependencies node itself and proceeding on to target framework, groupings, and then the individual top-level dependencies. Along the way it incrementally generates new `IProjectTree`s as it finds dependencies that have been updated, added, or removed.
```

Comment:
One important thing to note here is that since we are updating tree nodes form top to bottom it prevents Solution explorer to flick (close already opened nodes during those updates).